shm/Store.cpp

```diff
@@ -1,33 +1,170 @@
+
+Store::Store(std::shared_ptr<Time>& time) {
+    time_ = time;
+    time->addObserver(this);
+    makeStock();
+}
+
+size_t Store::genRand(int min, int max) const {
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<> distirb(min, max);
+
+    return static_cast<size_t>(distirb(gen));
+}
+
+void Store::generateFruits() {
+    size_t i = 0;
+    while(i < marketSection) {
+        Fruit fruit(fruitNames[genRand(0, 5)], genRand(1, 20), genRand(10, 30), genRand(1, 10));
+        if(std::none_of(begin(market_), end(market_),[&fruit](const auto& ptr){ return ptr->getName() == fruit.getName();})) {
+            market_.emplace_back(std::make_shared<Fruit>(fruit));
+            i++;
+        }
+    }
+}
+
+void Store::generateAlcos() {
+    size_t i = 0;
+    while(i < marketSection) {
+        Alcohol alco(alcoNames[genRand(0, 5)], genRand(1, 5), genRand(30, 100), genRand(40, 96));
+        if(std::none_of(begin(market_), end(market_),[&alco](const auto& ptr){ return ptr->getName() == alco.getName();})) {
+            market_.emplace_back(std::make_shared<Alcohol>(alco));
+            i++;
+        }
+    }
+}
+
+void Store::generateItems() {
+    size_t i = 0;
+    while(i < marketSection) {
+        Item item(itemNames[genRand(0, 5)], genRand(1, 10), genRand(30, 100), possibleRarities[genRand(0, 3)]);
+        if(std::none_of(begin(market_), end(market_),[&item](const auto& ptr){ return ptr->getName() == item.getName();})) {
+            market_.emplace_back(std::make_shared<Item>(item));
+            i++;
+        }
+    }
+}
+
+void Store::makeStock() {
+    market_.reserve(marketSection * 3);
+    generateFruits();
+    generateAlcos();
+    generateItems();
+    market_.shrink_to_fit();
+}
+
+CargoPtr Store::makeNewCargo(const CargoPtr& cargo, size_t amount) const {
+    if (typeid(*cargo) == typeid(Fruit)) {
+        Fruit fruit(cargo->getName(), amount, cargo->getBasePrice(), cargo->getUniqueStat());
+        return std::make_shared<Fruit>(fruit);
+    } else if (typeid(*cargo) == typeid(Alcohol)) {
+        Alcohol alco(cargo->getName(), amount, cargo->getBasePrice(), cargo->getUniqueStat());
+        return std::make_shared<Alcohol>(alco);
+    } else if (typeid(*cargo) == typeid(Item)) {
+        Item item(cargo->getName(), amount, cargo->getBasePrice(), static_cast<Rarity>(cargo->getUniqueStat()));
+        return std::make_shared<Item>(item);
+    }
+
+    return nullptr;
+}
+
+CargoPtr Store::getCargo(size_t index) const {
+    return market_[index - 1];
 }
 
-Response Store::buy(Cargo* cargo, size_t amount, Player* player) {
-    const size_t price = amount * cargo->getBasePrice();
+void Store::removeFromStore(const CargoPtr& cargo, size_t amount) {
+    if (amount == cargo->getAmount()) {
+        market_.erase(std::remove(begin(market_), end(market_), cargo), market_.end());
+    } else {
+        *cargo -= amount;
+    }
+}
```

Comment:
Pretty much the same code as `unload()` in Ship class. Maybe we can refactor it? Similarly to what I wrote about `findCargo` method.